### ISSUE TITLE
CONSTRUCT DISTINCT & REDUCED

### ISSUE THREAD

**Issue by RickMoynihan** [w3c/sparql-12#86]
## Why?

There is no easy way to force a triplestore to dedupe graph results before sending them to you over the wire.

Some construct queries can lead to an explosion of materialised (and duplicated) triples, especially where the engine has performed several cartesian joins.  For example a case I ran into recently resulted in almost 600,000 duplicated triples being sent over the wire where the number of distinct triples was just 63.

You can work around this by performing multiple queries, but this adds complexity to the application.  A `CONSTRUCT DISTINCT` would allow applications to fetch whole graphs of data in one shot, where the total amount of data is relatively small.

See also issue #67 

## Previous work

?

## Proposed solution

```sparql
CONSTRUCT DISTINCT {
   :foo ?bar ?baz .
   ?foo ?foop ?fooo .
   ?bar ?barp ?baro .
   ?baz ?bazp ?bazo .
} WHERE { 
   :foo ?bar ?baz .
   ?foo ?foop ?fooo .
   ?bar ?barp ?baro .
   ?baz ?bazp ?bazo .
} 
```

## Considerations for backward compatibility

None that I can think of.  An additive change to the grammar.

**Comment by kasei** [comment]
Can you discuss what implementations you use that produce redundant triples in the output? While the RDF serialization is probably valid, my reading of [SPARQL 1.1 regarding `CONSTRUCT`](https://www.w3.org/TR/sparql11-query/#construct) suggests that producing duplicate triples is probably against the spec:

> The result is an RDF graph formed by taking each query solution in the solution sequence, substituting for the variables in the graph template, and combining the triples into a single RDF graph by *set union*.

(emphasis added)

**Comment by kasei** [comment]
@cygri I agree with all of that. My point is that the text of the SPARQL spec says that the graph is constructed by set union. Streaming solutions that emit duplicate triples produce semantically equivalent RDF documents, but I don't see how you can do that and say you've done it by set union.

**Comment by RickMoynihan** [comment]
@kasei Thanks for the spec pointer.  I'm currently using stardog, though I'd expect most triplestores not to engage in perfect deduping for the reasons @cygri states.  Otherwise they'll struggle to stream results with constant space.

One thing the spec tends to be very vague on is who is doing what operation, or even when the triples are serialised/deserialised etc; so this might provide a loop-hole for stardog and other implementations, as they can pragmatically delay the de-duping to read-time.

> ... The result is an RDF graph formed by taking each query solution in the solution sequence, substituting for the variables in the graph template, and combining the triples into a single RDF graph by set union.

Given what you've highlighted though I think then there's an argument that 1.2 could remove and relax the "by set union" constraint in SPARQL 1.2 and enforce that explicitly as part of a `CONSTRUCT DISTINCT` operation.

**Comment by cygri** [comment]
@kasei But I don't have to do it by set union. The spec only defines what the correct result is. It doesn't constrain how a processor is to arrive at that result. If a processor finds a shortcut that still produces the correct result, then the processor conforms. Outputting the correct graph serialised with redundancy is just such a shortcut.

**Comment by kasei** [comment]
@cygri that's a fair point. I guess it depends on whether you think the "correct result" here is the RDF graph as read on the client side, or the (possibly serialized form of the) graph on the server side. I think for implementation convenience I want to side with you on the former, but my pedantic reading leans towards the latter. Perhaps more than the proposal above, we can work on clarifying the current language (or, as suggested above, explicitly relaxing the semantics while providing an explicit `DISTINCT` variant) and documenting best practices for dealing with large results.

**Comment by afs** [comment]
What about using a nested `SELECT`?

    CONSTRUCT { template }
    WHERE {
      SELECT REDUCED ?x1 ?2 {
        pattern
      }
    }

**Comment by afs** [comment]
Personally, I think the "set union" language should be put into the errata. RDF syntaxes don't require uniqueness of triples and the use case for streaming is strong. The client can impose the uniqueness condition at no extra space other than the space for the graph, or even issue a `SELECT` and apply the template locally, and engines could respect the nested `SELECT`.

See also issue #33 "Allow CONSTRUCT subqueries in FROM clauses" where the set semantics will be visible because a query can count the inner `CONSTRUCT`.

**Comment by cygri** [comment]
The problem I have with the explicit `CONSTRUCT DISTINCT` is that it is really about formatting the result in a particular way, and not about getting a different result. (Because removing duplicate triples do not change the meaning of an RDF document.) It is kind of similar to wanting the result pretty-printed or not, or wanting a particular JSON-LD context or frame applied.

My instinct is that SPARQL language features should be concerned with asking for a particular result, and not with asking for a specific presentation of that result.

Thinking out loud:

- Tarql has a `--dedup n` command-line argument that deduplicates CONSTRUCTed triples within a sliding window as they are streamed.
- The desire to have duplicates removed could be communicated with a `&dedup=n` or `&dedup=true` option in the SPARQL protocol.
- #64 asks for a syntax for named arguments for functions. Maybe that syntax can be re-used to provide a space where this kind of option can be communicated in the query, without specifying what the options mean?
    ```sparql
    PREFIX ...
    OPTIONS (_dedup=true, pretty-print=true)
    CONSTRUCT { ... }
    ```
    So the spec would say that `OPTIONS` is an extension point where processors can receive additional information that doesn't change the semantics of the query. This could potentially also be used for things like timeout, query hints, and so on.

**Comment by RickMoynihan** [comment]
> What about using a nested SELECT?

@afs I'm not sure this works as `SELECT DISTINCT`\`SELECT REDUCED` eliminate duplicate rows, i.e. where the whole row in the emitted solution is identical.  The majority of the duplicates I see in practice are not caused by duplicated solutions, but are duplicates within the solutions, themselves... e.g. where `?column1 `?column2` and `?column3` are the same but `?columns` 5, and 6 are different.

Usually these duplicates are the product of a crossjoin within the query.

**Comment by RickMoynihan** [comment]
> See also issue #33 "Allow CONSTRUCT subqueries in FROM clauses" where the set semantics will be visible because a query can count the inner CONSTRUCT.

@afs Thanks for referencing issue #33.  I suspect it might allow you to build a `CONSTRUCT DISTINCT`, not sure if this syntax is as proposed in that issue, but you could imagine something like might do what I want in a slightly verbose way...

```sparql
CONSTRUCT { ?s ?p ?o } WHERE {
  SELECT DISTINCT { ?s ?p ?o } FROM {
    UNION { 
      CONSTRUCT { 
        ?sc ?scp ?sco .
      }   
      WHERE {
        ?sc a skos:ConceptScheme .
        ?sc ?scp ?sco .
      }
    } {
      CONSTRUCT {
        ?co ?cop ?cos .
      } WHERE {
        ?sc a skos:ConceptScheme .
        ?co skos:inScheme ?sc .
        ?co ?cop ?cos .
      }
    }
  } WHERE {
    ?s ?p ?o .
  }
}
```

**Comment by TallTed** [comment]
(github seems to be having timestamp issues. the last two comments above this one are currently showing as having been made/edited "5 hours from now", i.e., those actions *will take place* but have not yet. Caution is urged for all activities impacted by timestamps.)

**Comment by RickMoynihan** [comment]
@afs yeah I'm actually coming to @cygri's view that this is more about formatting the result, and probably shouldn't be part of the standard.

Cross joins will I think naturally lead to something like N! duplicates, where N is the number of joins; though if cross joins were implemented with REDUCED like features they could kill most of those dupes and deduplicate in N space.

So I think this is really a deficiency in the implementation rather than a feature for 1.2; though I would suggest clarifying this in the specification/errata for 1.2.

**Comment by TallTed** [comment]
I wonder if `CONSTRUCT REDUCED` would be sufficient?  Certainly easier to implement on the server, whether its caching or streaming the result set, if it's just checking "does this triple duplicate _one of the last n_ I sent/cached" (with some moderate limits on the size of _n_) vs "does this triple duplicate _any_ I've already sent/cached".

**Comment by VladimirAlexiev** [comment]
@rickMoynihan may i suggest you rewrite your query not to do 10k times Cartesian explosion? use union for every multi value field.

Your request is akin to "need special trucks to clean up after nuclear explosion"
